pkg/solana/logpoller/loader.go

```diff
@@ -279,3 +303,191 @@ func (c *EncodedLogCollector) loadSlotBlocksRange(ctx context.Context, start, en
+	p.mu.Lock()
+	defer p.mu.Unlock()
+
+	p.blocks = append(p.blocks, block)
+}
+
+func (p *orderedParser) ExpectTxs(block uint64, quantity int) {
+	p.mu.Lock()
+	defer p.mu.Unlock()
+
+	p.expect[block] = quantity
+	p.actual[block] = make([]ProgramEvent, 0, quantity)
+}
+
+func (p *orderedParser) Process(event ProgramEvent) error {
+	p.mu.Lock()
+	defer p.mu.Unlock()
+
+	if err := p.addToExpectations(event); err != nil {
+		// TODO: log error because this is an unrecoverable error
+		return nil
+	}
+
+	return p.sendReadySlots()
+}
+
+func (p *orderedParser) start(_ context.Context) error {
+	p.engine.GoTick(services.NewTicker(time.Second), p.run)
+
+	return nil
+}
+
+func (p *orderedParser) close() error {
+	return nil
+}
+
+func (p *orderedParser) addToExpectations(evt ProgramEvent) error {
+	expectations, ok := p.expect[evt.SlotNumber]
+	if !ok {
+		return fmt.Errorf("%w: %d", errExpectationsNotSet, evt.SlotNumber)
+	}
+
+	evts, ok := p.actual[evt.SlotNumber]
+	if !ok {
+		return fmt.Errorf("%w: %d", errExpectationsNotSet, evt.SlotNumber)
+	}
+
+	p.actual[evt.SlotNumber] = append(evts, evt)
+
+	if expectations == len(evts)+1 {
+		p.setReady(evt.SlotNumber)
+	}
+
+	return nil
+}
+
+func (p *orderedParser) expectations(block uint64) (int, bool, error) {
+	expectations, ok := p.expect[block]
+	if !ok {
+		return 0, false, fmt.Errorf("%w: %d", errExpectationsNotSet, block)
+	}
+
+	evts, ok := p.actual[block]
+	if !ok {
+		return 0, false, fmt.Errorf("%w: %d", errExpectationsNotSet, block)
+	}
+
+	return expectations, expectations == len(evts), nil
+}
+
+func (p *orderedParser) clearExpectations(block uint64) {
+	delete(p.expect, block)
+	delete(p.actual, block)
+}
+
+func (p *orderedParser) setReady(slot uint64) {
+	p.ready = append(p.ready, slot)
+}
+
+func (p *orderedParser) run(_ context.Context) {
+	p.mu.Lock()
+	defer p.mu.Unlock()
+
+	_ = p.sendReadySlots()
+}
+
+func (p *orderedParser) sendReadySlots() error {
+	// start at the lowest block and find ready blocks
+	for _, block := range p.blocks {
+		// if no expectations are set, we are still waiting on information for the block.
+		// if expectations set and not met, we are still waiting on information for the block
+		// no other block data should be sent until this is resolved
+		exp, met, err := p.expectations(block)
+		if err != nil || !met {
+			break
+		}
+
+		// if expectations are 0 -> remove and continue
+		if exp == 0 {
+			p.clearExpectations(block)
+			p.blocks = p.blocks[1:]
+
+			continue
+		}
+
+		// to ensure ordered delivery, break from the loop if a ready block isn't found
+		rIdx := slices.Index(p.ready, block)
+		if rIdx < 0 {
+			return nil
+		}
+
+		evts, ok := p.actual[block]
+		if !ok {
+			return errInvalidState
+		}
+
+		var errs error
+		for _, evt := range evts {
+			errs = errors.Join(errs, p.parser.Process(evt))
+		}
+
+		// need possible retry
+		if errs != nil {
+			return errs
+		}
+
+		p.ready = slices.Delete(p.ready, rIdx, rIdx+1)
+		p.blocks = p.blocks[1:]
```

Comment:
Seems like this is a memory leak. We never rest `p.blocks`

Reply:
Why do we need to reset `p.blocks`? This simply removes the first element.

Reply:
If this were an array of pointers or slices, it would be a memory leak without doing `p.blocks[0] = nil`. But at least based on this answer, it should not be an issue because it's just an array of primitives:
https://stackoverflow.com/questions/55045402/memory-leak-in-golang-slice

Reply:
Actually, maybe that answer is wrong? there's an example of a memory leak here with `[]byte` that does seem similar to what we're doing here:
https://go.dev/blog/slices-intro

Reply:
I asked @jmank88 about this and he said it could be a memory leak like @dhaidashenko says if used incorrectly. However, in this case since `append` is being used to add values, Go _should_ resize the underlying array correctly. To be safe, since we don't know the magic underneath for certain in this circumstance *and* the array should remain small anyway, I will use the linked list.